### PR TITLE
Apply line clamp (truncation) to page title

### LINKED ISSUE #655: [Enhancement] Truncate page-title if title is longer then your defined max-height of page-title
**Is your enhancement request related to a problem? Please describe.**
We would like to truncate our page-title when we have longer titles that break our truncation rule (scale from h1 to h3). 

In the page-template we want to define what our max-height of our page-titles are, and we want to scale font down before we truncate.

The text-scaling task can be found here: https://github.com/kirbydesign/designsystem/issues/481

So the rule should be triggered when there is no more space in the title and the font-size has gone down to h3.

**See also**

Left to right: 
H1 - H2 - H3 & truncate
Skærmbillede 2019-10-07 kl  19 50 07

### PR DESCRIPTION
**Pull request by jkaltoft**
This PR closes #655 (reference issue number here)

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [x] Enhancement (to existing content)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] cookbook application / infrastructure changes
- [ ] Other... Please describe:

## Which issue documents the current behaviour?

Issue Number: #655

## What is the new behavior?

Page title gets truncated by utilizing non-standard but widely supported CSS line-clamp. For non line-clamp support a fallback has been provided.

<table>
  <tr>
    <th>Without line-clamp</th>
    <th>With line-clamp</th>
    <th>Fallback</th>
  </tr>
  <tr>
    <td>

without-line-clamp

</td>
    <td>

line-clamp

</td>
    <td>

line-clamp-fallback

</td>
  </tr>
</table>

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: libs/designsystem/src/lib/helpers/line-clamp-helper.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class LineClampHelper {
+  setLineClamp(el: HTMLElement, maxLines: number, lineHeight: string): void {
+    el.style.setProperty('--line-clamp', `${maxLines}`);
+    el.style.setProperty('--line-height', lineHeight);
+    el.classList.add('kirby-line-clamp');
+  }
+}
```
Comment (jakobe) [inline libs/designsystem/src/lib/helpers/line-clamp-helper.ts:10]: Maybe a bit bloated having a dedicated provider for just setting 3 lines of css? 🤔

File: libs/designsystem/src/lib/helpers/line-clamp-helper.ts [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class LineClampHelper {
+  setLineClamp(el: HTMLElement, maxLines: number, lineHeight: string): void {
+    el.style.setProperty('--line-clamp', `${maxLines}`);
+    el.style.setProperty('--line-height', lineHeight);
+    el.classList.add('kirby-line-clamp');
```
Comment (jakobe) [inline libs/designsystem/src/lib/helpers/line-clamp-helper.ts:8]: We generally use Renderer2 for modifying the DOM...

File: libs/designsystem/src/lib/scss/base/_line-clamp.scss [hunk 3]
```diff
@@ -0,0 +1,51 @@
+.kirby-line-clamp {
+  line-height: var(--line-height);
+  overflow: hidden;
+
+  // Fallback when -webkit-line-clamp is not supported
```
Comment (jakobe) [inline libs/designsystem/src/lib/scss/base/_line-clamp.scss:5]: Consider wrapping in:
`@supports not (-webkit-line-clamp: 1) {`

File: apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts [hunk 4]
```diff
@@ -10,18 +10,27 @@ const config = {
...
-    <ng-container *ngTemplateOutlet="customTitle"></ng-container>
+    <ng-container *ngTemplateOutlet="toolbarTitle"></ng-container>
   </ng-template>
 
-  <ng-template #customTitle>
-    <div style="display: flex;">
-      <div style="overflow: hidden; text-overflow: ellipsis;">
-        Custom Title With a very long name
+  <ng-template #pageTitle>
+    <div style="display: flex; gap: 0.5rem;">
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts:22]: Consider moving inlined styles to comp's stylesheet

File: apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts [hunk 5]
```diff
@@ -10,18 +10,27 @@ const config = {
...
+      <h1 [kirbyFitHeading]="{ maxLines: 2 }">
+        Custom Titles with very long names that span multiple lines will be truncated
+      </h1>
+      <kirby-icon name="arrow-down"></kirby-icon>
+    </div>
+  </ng-template>
+
+  <ng-template #toolbarTitle>
+    <div style="display: flex; gap: 0.5rem;">
```
Comment (jakobe) [inline apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts:31]: Consider moving inlined styles to comp's stylesheet

**Review by jakobe — CHANGES_REQUESTED** [review]

File: libs/designsystem/src/lib/helpers/line-clamp-helper.ts (hunk 1 above)
Comment (jkaltoft) [reply libs/designsystem/src/lib/helpers/line-clamp-helper.ts:10]: Yeah, perhaps. We discussed it when the PR was in draft. I want to keep line clamp separate from the auto-resize-text functionality in the fit-headings directive. Do you have a suggestion for an alternative?

File: libs/designsystem/src/lib/helpers/line-clamp-helper.ts (hunk 2 above)
Comment (jkaltoft) [reply libs/designsystem/src/lib/helpers/line-clamp-helper.ts:8]: Renderer2 didn't apply the CSS custom properties, but afterwards I was told it should be possible, if an extra parameter is applied. Haven't looked into it yet.

Could this be the reason that it doesn't initially get applied in Safari?

File: apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts (hunk 4 above)
Comment (jkaltoft) [reply apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts:22]: I don't think there is a stylesheet for this example component. The inline styles were already there (except for the gap I added).

File: apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts (hunk 5 above)
Comment (jkaltoft) [reply apps/cookbook/src/app/examples/page-example/advanced/page-custom-title-example.component.ts:31]: 👆🏻

File: libs/designsystem/src/lib/helpers/line-clamp-helper.ts (hunk 2 above)
Comment (jkaltoft) [reply libs/designsystem/src/lib/helpers/line-clamp-helper.ts:8]: Changed it to use Renderer2

**Comment by jkaltoft** [comment]
Reviewer, please focus on (aside from regular review stuff):

- shouldScale() issue
- line-clamp-helper as a provider
- Changes to templates in page-custom-title-example.component.ts

**Review by hafstad — APPROVED** [review]

**Review by jakobe — APPROVED** [review]

**Pull request merged**